### PR TITLE
chore(drive): state transition observability

### PR DESCRIPTION
**Pull request by shumkov**
## Issue being fixed or feature implemented

We need more transparency on state transition validation and execution process. Also, we need to add additional params to filter logs and remove extra messages to reduce noise.

## What was done?

- Added logs for check_tx
- Refactored StateTransitionContainer to preserve the original order of state transition so we can log execution results in the same order as txs in block
- Provided ordered single message logs for state transition processing 
- Added additional labels to check_tx metrics
- Added metrics to state transition processing
- Fixed reverted state transition hex

## How Has This Been Tested?

None

## Breaking Changes

None

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [x] I have made corresponding changes to the documentation if needed

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: packages/rs-drive-abci/src/abci/handler/check_tx.rs [hunk 1]
```diff
@@ -53,6 +56,27 @@ where
                 .cloned()
                 .unwrap_or_default();
 
+            let state_transition_name = check_tx_result.state_transition_name.to_owned();
+
+            let message = match (r#type, code) {
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/src/abci/handler/check_tx.rs:61]: We use the info to pass serialized consensus error to client. This logic is for logging and metrics.

**Review by lklimek — APPROVED** [review]

File: packages/rs-drive-abci/src/execution/check_tx/v0/mod.rs [hunk 2]
```diff
@@ -92,6 +94,8 @@ where
         platform_state: &PlatformState,
         platform_version: &PlatformVersion,
     ) -> Result<ValidationResult<CheckTxResult, ConsensusError>, Error> {
+        let state_transition_hash = hashes::sha256::Hash::hash(raw_tx).to_byte_array();
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/check_tx/v0/mod.rs:97]: We don't want to do cryptographic hash on every check tx.

**Review by QuantumExplorer — CHANGES_REQUESTED** [review]

**Review by pauldelucia — DISMISSED** [review]
Very nice

**Review by pauldelucia — DISMISSED** [review]

**Review by QuantumExplorer — CHANGES_REQUESTED** [review]

File: packages/rs-dpp/src/errors/consensus/basic/decode/serialized_object_parsing_error.rs [hunk 3]
```diff
@@ -6,7 +6,7 @@ use platform_serialization_derive::{PlatformDeserialize, PlatformSerialize};
 use thiserror::Error;
 
 #[derive(Error, Debug, Clone, Encode, Decode, PlatformSerialize, PlatformDeserialize)]
-#[error("Parsing of serialized object failed due to: {parsing_error}")]
+#[error("State transition decoding failed: {parsing_error}")]
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-dpp/src/errors/consensus/basic/decode/serialized_object_parsing_error.rs:9]: ok

File: packages/rs-drive-abci/src/abci/handler/check_tx.rs [hunk 4]
```diff
@@ -61,20 +66,79 @@ where
...
+                        "removed from mempool with code {code} after re-check due to error: {}",
+                        first_consensus_error
+                            .expect("consensus error must be present with non-zero error code")
+                    ),
+                    _ => unreachable!("we have only 2 modes of check tx"),
+                };
+
+                let state_transition_hash = check_tx_result
+                    .state_transition_hash
+                    .expect("state transition hash must be present if trace level is enabled");
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/abci/handler/check_tx.rs:95]: No expects please as code can change and then we get a panic.

File: packages/rs-drive-abci/src/abci/handler/check_tx.rs [hunk 5]
```diff
@@ -61,20 +66,79 @@ where
...
+                        "rejected with code {code} due to error: {}",
+                        first_consensus_error
+                            .expect("consensus error must be present with non-zero error code")
+                    ),
+                    (1, _) => format!(
+                        "removed from mempool with code {code} after re-check due to error: {}",
+                        first_consensus_error
+                            .expect("consensus error must be present with non-zero error code")
+                    ),
+                    _ => unreachable!("we have only 2 modes of check tx"),
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/abci/handler/check_tx.rs:90]: Please return an error instead.

File: packages/rs-drive-abci/src/abci/handler/check_tx.rs [hunk 6]
```diff
@@ -61,20 +66,79 @@ where
...
+            let priority = check_tx_result.priority as i64;
+
+            if tracing::enabled!(tracing::Level::TRACE) {
+                let message = match (r#type, code) {
+                    (0, 0) => "added to mempool".to_string(),
+                    (1, 0) => "kept in mempool after re-check".to_string(),
+                    (0, _) => format!(
+                        "rejected with code {code} due to error: {}",
+                        first_consensus_error
+                            .expect("consensus error must be present with non-zero error code")
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/abci/handler/check_tx.rs:83]: No expects please as code can change and then we get a panic.

File: packages/rs-drive-abci/src/abci/handler/check_tx.rs [hunk 7]
```diff
@@ -61,20 +66,79 @@ where
...
+                    ),
+                    (1, _) => format!(
+                        "removed from mempool with code {code} after re-check due to error: {}",
+                        first_consensus_error
+                            .expect("consensus error must be present with non-zero error code")
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/abci/handler/check_tx.rs:88]: No expects please as code can change and then we get a panic.

File: packages/rs-drive-abci/src/execution/types/state_transition_container/v0/mod.rs [hunk 8]
```diff
@@ -1,58 +1,70 @@
...
-    pub(super) invalid_state_transitions_with_protocol_error: Vec<(&'a [u8], ProtocolError)>,
+#[derive(Debug)]
+pub struct InvalidEncodedStateTransition<'a> {
+    pub raw: &'a [u8],
+    pub error: ConsensusError,
+    pub elapsed_time: Duration,
+}
+
+#[derive(Debug)]
+pub struct FaultyStateTransition<'a> {
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/types/state_transition_container/v0/mod.rs:21]: Stop renaming things. This one is InvalidStateTransitionWithProtocolError.

File: packages/rs-drive-abci/src/execution/types/state_transition_container/v0/mod.rs [hunk 9]
```diff
@@ -1,58 +1,70 @@
...
-#[derive(Debug, Default)]
-pub struct StateTransitionContainerV0<'a> {
-    /// The asset lock state transitions
-    pub(super) valid_state_transitions: Vec<(&'a [u8], StateTransition)>,
-    /// Deserialization errors
-    pub(super) invalid_state_transitions: Vec<(&'a [u8], ConsensusError)>,
-    /// Deserialization errors that broke platform, these should not exist, but are still handled
-    pub(super) invalid_state_transitions_with_protocol_error: Vec<(&'a [u8], ProtocolError)>,
+#[derive(Debug)]
+pub struct InvalidEncodedStateTransition<'a> {
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/types/state_transition_container/v0/mod.rs:14]: Stop renaming things this one is InvalidStateTransition

File: packages/rs-drive-abci/src/execution/types/state_transition_container/v0/mod.rs [hunk 10]
```diff
@@ -1,58 +1,70 @@
...
+
+#[derive(Debug)]
+pub struct FaultyStateTransition<'a> {
+    pub raw: &'a [u8],
+    pub error: ProtocolError,
+    pub elapsed_time: Duration,
+}
+
+#[derive(Debug)]
+pub struct SuccessfullyDecodedStateTransition<'a> {
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/types/state_transition_container/v0/mod.rs:28]: This one is ValidStateTransition.

File: packages/rs-drive-abci/src/execution/types/state_transition_container/v0/mod.rs [hunk 11]
```diff
@@ -1,58 +1,70 @@
...
-    fn destructure(
-        self,
-    ) -> (
-        Vec<(&'a [u8], StateTransition)>,
-        Vec<(&'a [u8], ConsensusError)>,
-        Vec<(&'a [u8], ProtocolError)>,
-    );
+#[derive(Debug)]
+pub struct StateTransitionContainerV0<'a> {
+    state_transitions: Vec<DecodedStateTransition<'a>>,
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/types/state_transition_container/v0/mod.rs:38]: This is a refactoring, and I'm not sure why it is needed, because it should not be needed for the purpose of this PR.

File: packages/rs-drive-abci/src/execution/platform_events/state_transition_processing/process_raw_state_transitions/v0/mod.rs [hunk 12]
```diff
@@ -68,197 +74,239 @@ where
...
-            valid_state_transitions,
-            invalid_state_transitions,
-            invalid_execution_abci_state_transitions,
-        ) = state_transition_container.destructure();
-
         let mut processing_result = StateTransitionsProcessingResult::default();
 
-        for (raw_state_transition, state_transition) in valid_state_transitions {
-            tracing::trace!(?state_transition, "Processing state transition");
+        for decoded_state_transition in state_transition_container.into_iter() {
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/platform_events/state_transition_processing/process_raw_state_transitions/v0/mod.rs:79]: There is no point to refactor here away from the destructure. If you want to do that you need to do it in it's own PR as it is refactoring. We used to do valid state transitions first, we can't at this point just refactor to change the logic without an explanation as to why this change is good.

**Review by QuantumExplorer — CHANGES_REQUESTED** [review]

File: packages/rs-drive-abci/src/execution/platform_events/state_transition_processing/process_raw_state_transitions/v0/mod.rs (hunk 12 above)
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/platform_events/state_transition_processing/process_raw_state_transitions/v0/mod.rs:79]: Explained it on the call. Agreed that we need additional comments in code and PR description

File: packages/rs-drive-abci/src/execution/types/state_transition_container/v0/mod.rs (hunk 10 above)
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/types/state_transition_container/v0/mod.rs:32]: Agreed to keep the name

**Review by QuantumExplorer — APPROVED** [review]
